pkg/controller/perconaservermongodb/version.go

```diff
@@ -41,15 +39,25 @@ func (r *ReconcilePerconaServerMongoDB) sheduleEnsureVersion(cr *api.PerconaServ
 
 	if ok {
 		log.Info(fmt.Sprintf("remove job %s because of new %s", schedule.CronShedule, cr.Spec.UpgradeOptions.Schedule))
-		r.deleteEnsureVersion(schedule.ID)
+		r.deleteEnsureVersion(cr, schedule.ID)
 	}
 
-	log.Info(fmt.Sprintf("add new job: %s", cr.Spec.UpgradeOptions.Schedule))
 	id, err := r.crons.crons.AddFunc(cr.Spec.UpgradeOptions.Schedule, func() {
-		r.statusMutex.Lock()
-		defer r.statusMutex.Unlock()
+		nn := types.NamespacedName{
+			Name:      cr.Name,
+			Namespace: cr.Namespace,
+		}
 
-		if !atomic.CompareAndSwapInt32(&r.updateSync, updateDone, updateWait) {
+		l, ok := r.lockers.Load(nn.String())
+		if !ok {
+			log.Error(nil, "failed to get lock for cluster", "cluster", nn.String())
+			return
+		}
```

Comment:
No need to load the lock on each run. Lock for the cluster remains the same and should be loaded once before closure creates:
```go
	nn := types.NamespacedName{
		Name:      cr.Name,
		Namespace: cr.Namespace,
	}

	l, ok := r.lockers.Load(nn.String())
	if !ok {
		return errors.Errorf("failed to get lock for cluster %v", nn.String())
	}

	id, err := r.crons.crons.AddFunc(cr.Spec.UpgradeOptions.Schedule, func() {
		l.statusMutex.Lock()
		defer l.statusMutex.Unlock()
...
```

Reply:
https://github.com/percona/percona-server-mongodb-operator/pull/472/commits/9a8a46cec1d2a2523f4f9cf9fe19fb31147d5ca4